app/[[...path]]/page.tsx

```diff
@@ -0,0 +1,12 @@
+import { createElement } from 'react';
+
+import router from '@/theme/_router';
+
+export default async function Page(props: { params: { path?: string[] } }) {
+    const { path = [] } = props.params;
+
+    const { match, Page: PageComponent } = await router(path);
+
+    // @ts-expect-error Server Component
+    return createElement(PageComponent, match);
+}
```

Comment:
So I've used this catch-all Next.js route to workaround the limitations of ServerComponents and reduce each theme boilerplate connecting the newsroom markup to the Next.js app instance.

The limitations with Next.js 13 server components I had to workaround are:

- layout components cannot access path parts below its level
   - root layout will never know which locale it's being rendered for; the only way to get the first `/[locale]` subpath value is to introduce a second level of layout (`app/[locale]/layout.tsx`)
   - there cannot be multiple dynamic subpath routes on the hierarchy level: it's not possible to have `/[locale]` and `/[slug]` subpaths with an additional matching logic -- deciding whether it is a localized homepage or a story page.


The additional routing layer provides some benefits:

- abstracts the generic catch-all route complexity by providing a human-friendly API in binding page modules and layouts to route params
- provides a way to additionally check the matched params against the API data. E.g.:
   - `/[locale]` route will match only if there is a language enabled for this locale in the newsroom
   - `/[slug]` route will match only if there is a story with this slug

- I wanted to avoid having `/[locale-or-slug]` Next.js as the only route with dynamic rendering (as it'll make things harder to understand) and just went full on the additional dynamic router layer, wiring theme-domain pages to the routes (see `theme/_router.ts`)

Reply:
> and based on the checks would rewrite to `/story/[slug]`?

I don't like this particular aspect of your suggestion. I'd not want to do internal rewriting magic, and instead better keep things directly mapped between the public URL and internal structure.

Also, I think it's important to keep in mind that requirements change over time and may require more conflicting routes later (or even now, for a custom theme) and with this approach, there is some spare flexibility remaining to withstand extensions in the future. I'd not want to tailor-make the routing logic so it can only work for the current fixed set of requirements.

But anyway, it's a great point that having parts of this unloaded to a middleware may help. I'll explore it. Thanks!

Reply:
> Also, I think it's important to keep in mind that requirements change over time and may require more conflicting routes later

One particular use-case that we've been talking about internally a lot: evergreen content pages (i.e., unlike time-based stories feed, these could be published and displayed as top-level website pages). This feature will most likely require the introduction of another sale-level dynamic route.